lib/semantics/rewrite-parse-tree.cc

```diff
@@ -105,6 +108,74 @@ bool RewriteMutator::Pre(parser::ExecutionPart &x) {
   return true;
 }
 
+static DeclTypeSpec *GetType(const parser::Name &x) {
+  if (x.symbol != nullptr) {
+    return x.symbol->GetType();
+  } else {
+    return nullptr;
+  }
+}
+static DeclTypeSpec *GetType(const parser::StructureComponent &x) {
+  return GetType(x.component);
+}
+static DeclTypeSpec *GetType(const parser::DataRef &x) {
+  return std::visit(
+      common::visitors{
+          [](const parser::Name &name) { return GetType(name); },
+          [](const common::Indirection<parser::StructureComponent> &sc) {
+            return GetType(sc.value());
+          },
+          [](const common::Indirection<parser::ArrayElement> &sc) {
+            return GetType(sc.value().base);
+          },
+          [](const common::Indirection<parser::CoindexedNamedObject> &ci) {
+            return GetType(ci.value().base);
+          },
+      },
+      x.u);
+}
+static DeclTypeSpec *GetType(const parser::Substring &x) {
+  return GetType(std::get<parser::DataRef>(x.t));
+}
+static DeclTypeSpec *GetType(const parser::Designator &x) {
+  return std::visit([](const auto &y) { return GetType(y); }, x.u);
+}
+static DeclTypeSpec *GetType(const parser::ProcComponentRef &x) {
+  return GetType(x.v.thing);
+}
+static DeclTypeSpec *GetType(const parser::ProcedureDesignator &x) {
+  return std::visit([](const auto &y) { return GetType(y); }, x.u);
+}
+static DeclTypeSpec *GetType(const parser::Call &x) {
+  return GetType(std::get<parser::ProcedureDesignator>(x.t));
+}
+static DeclTypeSpec *GetType(const parser::FunctionReference &x) {
+  return GetType(x.v);
+}
+static DeclTypeSpec *GetType(const parser::Variable &x) {
+  return std::visit(
+      [](const auto &indirection) { return GetType(indirection.value()); },
+      x.u);
+}
```

Comment:
This set of GetType functions is similar to GetSimpleName that I recently added to tools.h. There the purpose is to start from an Expr or Variable and return a Name. (I don't think all the necessary overloadings are there yet.)

Also, in a change I'm working on I want to do something similar but get the .source data member from the node that has it, going through wrapper classes as well as from Variable into Designator, Call, etc.

I'm wondering if there is a way to have a shared implementation for this sort of thing.

Reply:
I'll put the guts of GetType into a new lib/parser/tools.h file.  If it obsoletes GetSimpleName, I'll combine them.

The code in expression analysis that descends into a tree to acquire source locations might be useful to you.